### ISSUE TITLE
The result of setting up methods in an hierarchy of interfaces depends on the order by which they're set up

### ISSUE THREAD

**Issue by dcastro** [moq/moq4#131]
**Test Types**:

``` csharp
public interface BaseA
{
    string Method();
}

public interface BaseB
{
    string Method();
}

public interface Derived : BaseA, BaseB
{
    new string Method();
}
```

**Reproduce bug:**

``` csharp
var mock = new Mock<Derived>();

mock.Setup(x => x.Method())
    .Returns("a");
mock.As<BaseA>().Setup(x => x.Method())
    .Returns("b");
mock.As<BaseB>().Setup(x => x.Method())
    .Returns("c");

var derived = mock.Object;
Assert.Equal("a", derived.Method());            //equals "c" <---- Test fails
Assert.Equal("b", (derived as BaseA).Method()); //equals "b"
Assert.Equal("c", (derived as BaseB).Method()); //equals "c"
```

If I change the order of the setups, so that `Derived.Method` is configured last, the program now prints "a", "b", "c" as expected and the tests pass.

``` csharp
var mock = new Mock<Derived>();

mock.As<BaseA>().Setup(x => x.Method())
    .Returns("b");
mock.As<BaseB>().Setup(x => x.Method())
    .Returns("c");
mock.Setup(x => x.Method())
    .Returns("a");

var derived = mock.Object;
Assert.Equal("a", derived.Method());            //equals "a"
Assert.Equal("b", (derived as BaseA).Method()); //equals "b"
Assert.Equal("c", (derived as BaseB).Method()); //equals "c"
```

**Comment by pjquirk** [comment]
I found this bug independently today working on an answer for [this question](http://stackoverflow.com/a/28090935/1698557) on StackOverflow.  Hopefully what I found can help.

**Comment by stakx** [comment]
See also #22, which is also about shadowed members. It appears that it's not entirely clear what Moq does / should do in such cases.

**Comment by stakx** [comment]
@pjquirk, well spotted, [this line of code](https://github.com/moq/moq4/blob/acd116563d5627a25e051ad1bdc0c231755533f0/Source/InterceptorStrategies.cs#L105), like you mentioned in your answer on Stack Overflow, indeed seems to be the reason for this peculiar behavior. I'll look into this tomorrow. My current guess is that it would be possible to first attempt to find a precise match, and only if there is none (or several equally good matches), fall back to the current `.LastOrDefault` behavior.

**Comment by stakx** [comment]
This should be fixed in the next release of Moq (version >4.7.63).

**Comment by pjquirk** [comment]
Thanks for the fix @stakx, and glad I could be of some use!
